### PR TITLE
Refactor distributed table creation logic

### PR DESCRIPTION
**Pull request by byucesoy**
This commit is preperation for introducing distributed partitioned table support.
We want to clean and refactor some code in distributed table creation logic so
that we can handle partitioned tables in more robust way.

### CODE REVIEW

File: src/backend/distributed/commands/create_distributed_table.c [hunk 1]
```diff
@@ -265,79 +273,201 @@ CreateReferenceTable(Oid relationId)
...
+	EnsureDistributionConfiguration(relationId, distributionColumn, distributionMethod,
+									colocationId, replicationModel, viaDeprecatedAPI);
+
+	/* we need to calculate these variables before creating distributed metadata */
+	localTableEmpty = LocalTableEmpty(relationId);
+	colocatedTableId = ColocatedTableId(colocationId);
+	if (colocatedTableId != InvalidOid)
+	{
+		/*
+		 * we take lock on colocatedTableId, because we want to ensure that colocated
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: Start with a capital

File: src/backend/distributed/commands/create_distributed_table.c [hunk 2]
```diff
@@ -265,79 +273,189 @@ CreateReferenceTable(Oid relationId)
...
+ */
+static char
+AppropriateReplicationModel(char distributionMethod, int replicationFactor,
+							bool viaDeprecatedAPI)
+{
+	if (viaDeprecatedAPI)
+	{
+		if (ReplicationModel != REPLICATION_MODEL_COORDINATOR)
+		{
+			ereport(NOTICE, (errmsg("using statement-based replication"),
```
Comment (aamederen) [inline src/backend/distributed/commands/create_distributed_table.c]: `ERROR` would exit the command here and we wanted to resume the operation but notice the user about the changes. I don't recall any specific reason for not erroring out.

Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: We already don't support this feature, so I'd rather error out on this. @byucesoy could you please change this to `ERROR`?

File: src/backend/distributed/commands/create_distributed_table.c [hunk 3]
```diff
@@ -265,79 +273,189 @@ CreateReferenceTable(Oid relationId)
...
-
-	/*
-	 * Lock target relation with an exclusive lock - there's no way to make
-	 * sense of this table until we've committed, and we don't want multiple
-	 * backends manipulating this relation.
-	 */
-	relation = relation_open(relationId, ExclusiveLock);
-	relationDesc = RelationGetDescr(relation);
-	relationName = RelationGetRelationName(relation);
+	relation = relation_open(relationId, AccessShareLock);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: And probably move down to `Ensure*` functions?

File: src/backend/distributed/commands/create_distributed_table.c [hunk 4]
```diff
@@ -146,199 +162,382 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
...
+	int replicationFactor = ShardReplicationFactor;
+	uint32 colocationId = INVALID_COLOCATION_ID;
+	Oid colocatedTableId = InvalidOid;
+	bool localTableEmpty = false;
+
+	Relation colocatedRelation = NULL;
+	Relation pgDistColocation = NULL;
+
+	/* generate colocation id */
+	if (distributionMethod == DISTRIBUTE_BY_APPEND ||
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: Hmm, why did we remove `ColocationIdForNewTable()`? That was hiding these details from `CreateDistributedTable()` which is fine.

I think if we can do, we could also move the following check into that function

```
	if (distributionMethod != DISTRIBUTE_BY_HASH &&
		colocateWithTableName != NULL &&
		pg_strncasecmp(colocateWithTableName, "default", NAMEDATALEN) != 0)
	{
		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
						errmsg("cannot distribute relation"),
						errdetail("Currently, colocate_with option is only supported "
								  "for hash distributed tables.")));
	}
```
And prevent sending `colocatedWithTableName` to `EnsureRelationCanBeDistributed()` function.

File: src/backend/distributed/commands/create_distributed_table.c [hunk 5]
```diff
@@ -146,199 +162,382 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
...
-	/* use configuration values for shard count and shard replication factor */
-	CreateHashDistributedTable(relationId, distributionColumnName,
-							   colocateWithTableName, ShardCount,
-							   ShardReplicationFactor);
+	/*
+	 * If we are using master_create_distributed_table, we don't need to continue,
+	 * because deprecated API does not supports the following features.
+	 */
+	if (viaDeprecatedAPI)
+	{
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: Maybe assert colocatedRelation == NULL ?

File: src/backend/distributed/commands/create_distributed_table.c [hunk 6]
```diff
@@ -146,199 +162,382 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
...
+	/* we need to calculate these variables before creating distributed metadata */
+	localTableEmpty = LocalTableEmpty(relationId);
+	colocatedTableId = ColocatedTableId(colocationId);
+	if (colocatedTableId != InvalidOid)
 	{
-		EnsureTableNotDistributed(relationId);
-		EnsureLocalTableEmpty(relationId);
+		/*
+		 * We take lock on colocatedTableId, because we want to ensure that colocated
+		 * table will be exist until we create all colocated shards.
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: `will be exist` maybe a better phrase?

File: src/backend/distributed/utils/colocation_utils.c [hunk 7]
```diff
@@ -904,6 +904,11 @@ ColocatedTableId(Oid colocationId)
 	ScanKeyData scanKey[1];
 	int scanKeyCount = 1;
 
+	if (colocationId == INVALID_COLOCATION_ID)
```
Comment (onderkalaci) [inline src/backend/distributed/utils/colocation_utils.c]: Single line comment

File: src/backend/distributed/commands/create_distributed_table.c [hunk 8]
```diff
@@ -146,199 +162,417 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
...
+ * with the given configuration. We perform almost all safety checks for distributing
+ * table here. If there is an unsatisfied requirement, we error out and do not
+ * distribute the table.
  *
- * XXX: We should perform more checks here to see if this table is fit for
- * partitioning. At a minimum, we should validate the following: (i) this node
- * runs as the master node, (ii) table does not make use of the inheritance
- * mechanism and (iii) table does not have collated columns.
+ * This function assumes, callers already took necessary locks to ensure there will not
+ * be any change in the given table.
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: ` given table.` -> `given relation` ?
`callers already ` - >`the callers have already acquired` ?

File: src/backend/distributed/commands/create_distributed_table.c [hunk 9]
```diff
@@ -146,199 +162,417 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
...
+/*
+ * CreateDistributedTable creates distributed table in the given configuration.
+ * This functions contains all necessary logic to create distributed tables. It
+ * perform necessary checks to ensure distributing the table is safe. If it is
+ * safe to distribute the table, this function creates distributed table metadata,
+ * creates shards and copies local data to shards.
+ *
+ * viaDeprecatedAPI boolean flag is not optimal way to implement this function,
+ * but it helps reducing code duplication a lot. We hope to remove that flag one
+ * day. once we deprecate master_create_distribute_table completely.
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: `.` -> `,`

File: src/backend/distributed/commands/create_distributed_table.c [hunk 10]
```diff
@@ -146,199 +162,417 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
...
-							   colocateWithTableName, ShardCount,
-							   ShardReplicationFactor);
-
 	if (ShouldSyncTableMetadata(relationId))
 	{
 		CreateTableMetadataOnWorkers(relationId);
 	}
 
-	PG_RETURN_VOID();
+	if (colocatedRelation != NULL)
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: I think this should happen just before `ShouldSyncTableMetadata ()`.

File: src/backend/distributed/commands/create_distributed_table.c [hunk 11]
```diff
@@ -146,199 +162,417 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
...
+	char replicationModel = REPLICATION_MODEL_INVALID;
+	uint32 colocationId = INVALID_COLOCATION_ID;
+	Oid colocatedTableId = InvalidOid;
+	bool localTableEmpty = false;
+
+	Relation colocatedRelation = NULL;
+
+	replicationModel = AppropriateReplicationModel(distributionMethod, viaDeprecatedAPI);
+
+	colocationId = ColocationIdForNewTable(relationId, distributionColumn,
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: `ColocationIdForNewTable ` assumes that the caller acquires the lock. But, the upper level caller acquired the lock in this case. Shall we add a single line comment saying this ?

File: src/backend/distributed/commands/create_distributed_table.c [hunk 12]
```diff
@@ -348,87 +582,233 @@ ConvertToDistributedTable(Oid relationId, char *distributionColumnName,
...
+			GetNodeUserDatabaseConnection(connectionFlag, nodeName, nodePort, NULL,
+										  NULL);
+
+		ExecuteCriticalRemoteCommand(connection, applySchemaCreationDDL->data);
+	}
+}
+
+
+/*
+ * In some cases, it is possible and safe to send local data to shards while
```
Comment (onderkalaci) [inline src/backend/distributed/commands/create_distributed_table.c]: Very minor: We generally start the comment with the function name?

**Pull request merged**